Web ImageFilter.matrix support

flar opened the pull request:
Includes a new golden test.

The only thing that doesn't work is using an ImageFilter.matrix in a BackdropFilter on the HTML renderer. I don't know of a way to apply a transform to the background.

What does work:
- ImageFiltered with matrix on both CanvasKit and HTML
- BackdropFilter with matrix on CanvasKit
- BackdropFilter does nothing with a matrix filter on HTML

Previously all of those would throw an exception.

flar commented:
Thanks @ferhatb , builds have been problematic for a while so I was just recently able to rebase this and get it past the build issues (hopefully, knocking on wood).

ferhatb commented on lib/web_ui/lib/src/engine/html/backdrop_filter.dart:
```diff
@@ -113,9 +113,9 @@ class PersistedBackdropFilter extends PersistedContainerSurface
+            filter._filterAttribute);
       }
-      DomRenderer.setElementStyle(_filterElement!, 'backdrop-filter', _imageFilterToCss(filter));
+      DomRenderer.setElementStyle(_filterElement!, 'backdrop-filter', filter._filterAttribute);
```
add space after filter,

flar replied:
I don't understand the comment here. What space?

flar replied:
Perhaps you read the punctuation after `filter` as a comma? It's a field reference...?

flar replied:
I'm going to resolve this conversation as I don't see anything wrong with the code. Feel free to re-comment if I missed something...

yjbanov commented on lib/web_ui/lib/src/engine/canvaskit/image_filter.dart:
```diff
@@ -109,3 +110,32 @@ class _CkBlurImageFilter extends CkImageFilter {
+  _CkMatrixImageFilter({ required this.matrix, required this.filterQuality }) : super._();
+
+  Float64List matrix;
+  ui.FilterQuality filterQuality;
```
Can these be `final`?

yjbanov replied:
Also, probably very minor performance wins, but if this object is opaque the values can be converted using `toSkMatrixFromFloat64` and `toSkFilterQuality` in the constructor initializer, so when the image filter is reused we don't need to re-convert.

flar replied:
At first I was worried that this would change the toString() representation, but then I realized I could convert back for stringifying, so I went with this method.

Also, I wasn't copying the matrix so that was a problem waiting to happen. By converting the matrix to an Sk form, we have a defensive copy implicitly (although the stringify precision will suffer slightly, but that is arguably being honest about the precision that web maintains for these values...).

flar replied:
In the end I made them final, but didn't store them as converted values in case that disrupts the processing of the == operator which is key to consistency of tree changes as values change.

yjbanov reviewed (approved):
![lgtm](https://user-images.githubusercontent.com/211513/32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d.png)

flar commented:
I implemented suggestions from @yjbanov and rebased. I had to make some adjustments for the recent changes by @hterkelsen for the way the libraries are packaged. Hopefully I complied with the new design directions.
